tutorials/3_transactions.md

```diff
@@ -99,127 +148,139 @@ struct Fund {
+its impact on completeness, are the topic of the next tutorial. For now we can
+think of this as a bounded model with two clients, so at the least, our
+verification is sound.
 
-Moving on, we enter a loop which simulates a sequence of transitions:
+At line 108 we then find the "transaction loop" which simulates a sequence of
+transitions:
 
 ```cpp
 smartace_log("[Entering transaction loop]");
 while (sol_continue()) {
     sol_on_transaction();
-    ((sender).v) = (rt_nd_range(3, 5, "sender"));
-    ((value).v) = (0);
-    ((blocknum_tmp).v) = (nd_uint256_t("blocknum"));
-    sol_require(((blocknum).v) <= ((blocknum_tmp).v), 0);
-    (blocknum) = (blocknum_tmp);
-    ((timestamp_tmp).v) = (nd_uint256_t("timestamp"));
-    sol_require(((timestamp).v) <= ((timestamp_tmp).v), 0);
-    (timestamp) = (timestamp_tmp);
-    uint8_t next_call = (rt_nd_range(0, 5, "next_call"));
-    sol_uint256_t pre = contract_1->model_balance;
+    /* ... Select non-deterministic blockchain state ... */
+    uint8_t next_call = nd_range(0, 5, "next_call");
     switch (next_call) {
     case 0: {
         smartace_log("[Calling openFund on contract_0]");
-        Method_Manager_FuncopenFund(&(contract_0), sender, value, blocknum, timestamp, paid, origin);
+        Manager_Method_openFund(&(contract_0), /* Blockchain state */);
         smartace_log("[Call successful]");
         break;
     }
-    /* ... */
+    /* ... Other methods ... */
     }
 }
 ```
 
-The `Seahorn` implementation of `sol_continue` always returns `true`, so the
-loop runs forever. On each iteration, a method is called non-deterministically.
-The method is determined by `next_call`. To instrument the contract, we must
-find the case which corresponds to `Manager.openFund()`. This is easy if we
-understand how SmartAce names functions:
+The methods `sol_continue()` and `sol_on_transaction()` are runtime hooks
+provided by `libverify/`. For `Seahorn`, `sol_continue()` is alway true while
+`sol_on_transaction` is a no-op. This means that the loop runs forever. On each
+iteration, a method is selected non-deterministically, and then executed with
+non-deterministic arguments. The simulation ends if an assertion fails.
 
-  * A constructor is an unnamed method, which is mapped to `Init_Manager`.
-  * A fallback is an unnamed method, which is mapped to `Fallback_Manager`.
-  * A method, say `openFund`, is mapped to `Method_Manager_FuncopenFund`.
+### Instrumenting the Model
 
-(*SW: This example doesn't talk about inheritance. If contract `A` extends `B` and calls `B.f` we write `Method_B_for_A_funcf`.*)
+We can encode the monitor using three ghost variables and a single assertion.
+We start by declaring each variable at the beginning of `cmodel.c`.
 
-(*SW: I suggest `CONTRACT_Method(_For_OVERRIDE)_METHODNAME`, `CONTRACT_INIT`, and `CONTRACT_FALLBACK`. Modifiers will come before `_METHODNAME`.*).
+```cpp
+GHOST_VAR int called_openFund;
```

Comment:
I prefer to name global variables with `g_` prefix, so this is `g_called_openFund`.
Here, nicely, `g_` stands both for **ghost** and **global**